browser/lib/transport/jsonptransport.ts

```diff
@@ -0,0 +1,232 @@
+			if(headers['X-Ably-Version']) this.params.v = headers['X-Ably-Version'];
+			if(headers['X-Ably-Lib']) this.params.lib = headers['X-Ably-Lib'];
+		}
+		this.body = body;
+		this.method = method;
+		this.requestMode = requestMode;
+		this.timeouts = timeouts;
+		this.requestComplete = false;
+	}
+
+	exec() {
+		const id = this.id,
+			body = this.body,
+			method = this.method,
+			uri = this.uri,
+			params = this.params;
+
+		params.callback = '_ablyjs_jsonp._(' + id + ')';
+
+		params.envelope = 'jsonp';
+		if(body) {
+			params.body = body as string;
+		}
+		if(method && method !== 'get') {
+			params.method = method;
+		}
+
+		const script = this.script = document.createElement('script');
+		const src = uri + Utils.toQueryString(params);
+		script.src = src;
+		if(script.src.split('/').slice(-1)[0] !== src.split('/').slice(-1)[0]) {
+			/* The src has been truncated. Can't abort, but can at least emit an
+			 * error so the user knows what's gone wrong. (Can't compare strings
+			 * directly as src may have a port, script.src won't) */
+			Logger.logAction(Logger.LOG_ERROR, 'JSONP Request.exec()', 'Warning: the browser appears to have truncated the script URI. This will likely result in the request failing due to an unparseable body param');
+		}
+		script.async = true;
+		script.type = 'text/javascript';
+		script.charset = 'UTF-8';
+		script.onerror = (err: string | Event) => {
+			this.complete(new ErrorInfo('JSONP script error (event: ' + Utils.inspect(err) + ')', null, 400));
+		};
+
+		type JSONPResponse = {
+			statusCode?: number;
+			response: {
+				error?: ErrorInfo;
+			};
+			headers?: Record<string, string>;
+		}
+
+		_['_' + id] = (message: JSONPResponse) => {
+			if(message.statusCode) {
+				/* Handle as enveloped jsonp, as all jsonp transport uses should be */
+				const response = message.response;
+				if(message.statusCode == 204) {
+					this.complete(null, null, null, message.statusCode);
+				} else if(!response) {
+					this.complete(new ErrorInfo('Invalid server response: no envelope detected', null, 500));
+				} else if(message.statusCode < 400 || Utils.isArray(response)) {
+					/* If response is an array, it's an array of protocol messages -- even if
+					 * it contains an error action (hence the nonsuccess statuscode), we can
+					 * consider the request to have succeeded, just pass it on to
+					 * onProtocolMessage to decide what to do */
+					this.complete(null, response, message.headers, message.statusCode);
+				} else {
+					const err = response.error || new ErrorInfo('Error response received from server', null, message.statusCode);
+					this.complete(err);
+				}
+			} else {
+				/* Handle as non-enveloped -- as will be eg from a customer's authUrl server */
+				this.complete(null, message);
+			}
+		};
+
+		const timeout = (this.requestMode == XHRStates.REQ_SEND) ? this.timeouts.httpRequestTimeout : this.timeouts.recvTimeout;
+		this.timer = setTimeout(() => { this.abort(); }, timeout);
```

Comment:
sorry I know this is a direct port from the js version but isn't this functionally the same as this:
```suggestion
		this.timer = setTimeout(this.abort.bind(this), timeout);
```